plugins/cwe_checker_emulation/cwe_checker_emulation.ml

```diff
@@ -0,0 +1,178 @@
+let entry_point_collector = object
+  inherit [tid list] Term.visitor
+  method! enter_term _ t entries =
+    if Term.has_attr t Sub.entry_point
+    then Term.tid t :: entries
+    else entries
+  end
+
+(** Wrapper function around entry_point_collector. *)
+let entry_points prog =
+  entry_point_collector#run prog []
+
+(** Collects all subroutines of the program that
+that are not an entry point. *)
+let all_subroutines prog =
+  let entries = entry_points prog in
+  let non_entry =
+    let roots = Tid.Set.of_list entries in
+    fun t -> if Set.mem roots t then None else Some (`tid t) in
+  List.map entries ~f:(fun t -> `tid t) @
+  Seq.to_list @@
+  Seq.filter_map ~f:non_entry @@
+  Graphlib.reverse_postorder_traverse (module Graphs.Callgraph) @@
+    Program.to_graph prog
+
+(** Executes/forks another Primus machine. *)
+let exec x =
+  Machine.current () >>= fun cid ->
+  info "Fork %a: starting from the %s entry point"
+    pp_id cid (string_of_name x);
+  Machine.catch (Linker.exec x)
+    (fun exn ->
+       info "execution from %s terminated with: %s "
+         (string_of_name x)
+         (Primus.Exn.to_string exn);
+       Machine.return ())
+
+let rec run = function
+  | [] ->
+    info "all toplevel machines done, halting";
+    Eval.halt >>=
+    never_returns
+  | x :: xs ->
+    Machine.current () >>= fun pid ->
+    Machine.fork ()    >>= fun () ->
+    Machine.current () >>= fun cid ->
+    if pid = cid
+    then run xs
+    else
+      exec x >>= fun () ->
+      Eval.halt >>=
+      never_returns
+
+(** Checks if a certain Primus.Observation.Provider is equal
+    to a string like 'incident'. *)
+let has_name name p =
+  Primus.Observation.Provider.name p = name
+
+(** Register a monitor. *)
+let monitor_provider name ps =
+  Primus.Observation.list_providers () |>
+  List.find ~f:(has_name name) |> function
+  | None -> invalid_argf "An unknown observation provider `%s'" name ()
+  | Some p -> p :: ps
+
+let parse_monitors =
+  List.fold ~init:[] ~f:(fun ps name -> monitor_provider name ps)
+
+(** Register monitors for 'incident' related events. *)
+module Monitor(Machine : Primus.Machine.S) = struct
+    open Machine.Syntax
+
+    let init () =
+      parse_monitors ["incident"; "incident-location"] |>
+      List.iter ~f:(fun m ->
+          info "monitoring %s" (Primus.Observation.Provider.name m);
+          Stream.observe (Primus.Observation.Provider.data m) (collect_events m));
+      Machine.return ()
+end
+
+(** Main logic of program:
+- we monitor all 'incident' related events
+- for all subroutins we fork a Primus machine
+- all monitored events are collected globally
+- after the last Primus machine has terminated we report all observed incidents *)
+let main {Config.get=(!)} proj =
+  Log_utils.set_log_level Log_utils.DEBUG;
+  Log_utils.set_output stdout;
+  Log_utils.color_on ();
+
+  Primus.Machine.add_component (module Monitor);
+  begin
+  let targets =  all_subroutines (Project.program proj) in
```

Comment:
Why are entry points filtered out from targets?

Reply:
The basic idea is to focus on a per subroutine emulation.

Reply:
Now we try to emulate all subroutines.